### PR TITLE
group routes

### PR DESCRIPTION
**Pull request by ghost**
-added jget routes for retrieving user's groups and a list of friends to add to a group
-added jpost routes to create a group and add a user that is a friend to a group 
-added test to UserTest to test User method getFriendsToAdd, which returns a list of friends to add to a group that is not already in that group
-testGetFriendsToAdd was failing b/c we were not handling the case where a user did not have any friends. now when a user does not have any friends or when all of a user's friends are in a group, we return an array list of size 0.
-added @Reference to list of users and list of groups

### CODE REVIEW

**Comment by jjalfaro9** [comment]
retry

File: src/main/java/com/johnstarich/moviematcher/models/User.java [hunk 1]
```diff
@@ -229,11 +233,31 @@ public User addFriendToGroup(String groupName, User newMember) throws HttpExcept
...
+		Optional<Group> g = groups.parallelStream().filter(group -> group.name.equals(groupName)).findFirst();
+		if(! g.isPresent()) { throw new HttpException(HttpStatus.BAD_REQUEST, "Could not find "+groupName); }
+		/** I want the users who are my friends and not in this group */
+		/** these are members of the group , so return friends who are not in the member */
+		if(friends == null) return Optional.of(new ArrayList<>(0));
+		if(g.get().members == null) return Optional.of(new ArrayList<>(friends));
+		User[] couldAdd =  friends
+			.parallelStream()
+			.filter( friend -> g.get().members.parallelStream().noneMatch(Predicate.isEqual(friend)))
+			.toArray(User[]::new);
```
Comment (JohnStarich) [inline src/main/java/com/johnstarich/moviematcher/models/User.java]: please replace with

``` java
return Optional.ofNullable(friends
            .parallelStream()
            .filter( friend -> g.get().members.parallelStream().noneMatch(Predicate.isEqual(friend)))
            .collect(Collectors.toList()))
```

**Comment by jjalfaro9** [comment]
retry

**Comment by JohnStarich** [comment]
LGTM 👍

File: src/main/java/com/johnstarich/moviematcher/models/User.java [hunk 2]
```diff
@@ -230,13 +234,18 @@ public User addFriendToGroup(String groupName, User newMember) throws HttpExcept
 		}
 
 		Group editThisGroup = editThisGroupOptional.get();
-
+		if(editThisGroup.members.contains(newMember)) { throw new HttpException(HttpStatus.BAD_REQUEST, newMember.username + " is already in group"); }
 		groups.remove(editThisGroup);
```
Comment (jjalfaro9) [inline src/main/java/com/johnstarich/moviematcher/models/User.java]: My bad Cesar, we need to make edits to this code. When I wrote this I made some major bugs ... sorry. We will look at it together.

**Comment by JohnStarich** [comment]
Some issues with how load is called and otherwise. Approving for now, but this introduces technical debt we need to fix for sure.

**Pull request merged**